Tooling updates + add Proxy support to httpclient.DefaultTransport

mmlb opened the pull request:
## What does this PR implement/change/remove?

Some spring cleaning of non-source files and addition of Proxy config to the httpclient.DefaultTransport.

Spring cleaning changes are annoyingly large but mostly mechanical, either via `gofumpt` or `golangci-lint run --fix`, there were some manual fixes I made to avoid disabling linters that complained about a handful of things.

I also added Proxy config to the DefaultTransport just like the DefaultTransport in the http package does. This setup will make it so http requests will go via a proxy when HTTPS_PROXY env var exists. This way we can debug redfish over https using mitmproxy or similar setup.

## Description for changelog/release notes

```
https connections now go through a proxy if the HTTPS_PROXY environment variable is set
```

jacobweinstock commented on internal/redfishwrapper/task.go:
```diff
@@ -76,13 +74,13 @@ func (c *Client) ConvertTaskState(state string) constants.TaskState {
 	case "starting", "downloading", "downloaded", "scheduling":
 		return constants.Initializing
-	case "running", "stopping", "cancelling":
+	case "running", "stopping", "canceling":
```
I don't think these should change as they are States that correspond to the gofish library. https://github.com/stmcginnis/gofish/blob/4b51c15f0a3df649821f9fc395a82483ff60bb2a/redfish/task.go#L16-L55 

It is spelled `Cancelling` in that library. We should probably be using the const from that package anyway.

mmlb replied:
uggh good catch will back out. We are using the consts from that package already. The problem is we are converting that package's types <-> string when it should be done by that package itself (via stringer imo).

mmlb replied:
should be back to original (plus `//nolint:` directive)

jacobweinstock commented on client.go:
```diff
@@ -148,12 +148,20 @@ func (c *Client) defaultTimeout(ctx context.Context) time.Duration {
+	client := *old
+	transport, ok := client.Transport.(*http.Transport)
+	if !ok {
+		panic("old client's transport is not expected type")
```
hmm...I'm concerned about this panic. I know we weren't even checking the type assert before, and it could have implicitly caused a panic already, but could we introduce a more graceful way to handle the bad type assert? Maybe a return error or something similar?

mmlb replied:
I don't think its worth it, changing to an error return is going to have wide effect when done for `registerASRRProvider` and probably others (bubbling the error up). Its never really going to be a problem either since we aren't really using different implementations. If it ever changes then it'll be caught early at startup instead of later at runtime.

Not a whole lot of benefit imo.

joelrebel replied:
This does seem like an exceptional case, can the panic message include a clearer message, along the lines of "expected a http client of the http.Transport type, got $blah instead"?

mmlb replied:
I've updated the panic message

mmlb commented on go.mod:
```diff
@@ -1,6 +1,6 @@
 module github.com/bmc-toolbox/bmclib/v2
 
-go 1.21
+go 1.23
```
Right, but no one should be running go <1.23 anyway. I kind of see it as a chicken/egg situation. When 1.25 comes out I'd like to bump this to 1.24 and then make use of `go tool` support to run various tools (like golangci-lint) via native go pinning.

jacobweinstock replied:
Sorry, that's not the [policy](https://github.com/bmc-toolbox/bmclib?tab=readme-ov-file#go-version-in-gomod). As a library, we shouldn't dictate the version clients use. Per the policy, `go tool` support isn't a reason to update this. There are many examples of older Go versions in large active Go module repos in the wild. https://github.com/spf13/cobra, for example.

mmlb replied:
oh til! well will back this out then

mmlb replied:
there's also the `copyloopvar` change that came in from 1.22 that affects this repo (albeit just one line). I'll back this out and have a separate PR for 1.22 + enabling those linters.

mmlb replied:
Alright dropped back down to 1.21 and undid changes

mmlb commented on providers/asrockrack/helpers.go:
```diff
@@ -602,7 +600,7 @@ func (a *ASRockRack) queryHTTPS(ctx context.Context, endpoint, method string, pa
 
 	// add headers
-	req.Header.Add("X-CSRFTOKEN", a.loginSession.CSRFToken)
+	req.Header.Add("X-Csrftoken", a.loginSession.CSRFToken)
```
this was an autofix from one of the linters, can do the magic comment to have it ignore this but I'd like some clarity before going that route.

I'm not seeing anything in the code base about old firmware expecting specific case of headers, are you thinking about supermicro with form-data fields https://github.com/bmc-toolbox/bmclib/blob/main/providers/supermicro/floppy.go#L91-L96?

joelrebel replied:
I'd recommend the headers are left as is, unless a change is required in practice

mmlb replied:
will do

mmlb replied:
should be back to original format

joelrebel reviewed (changes requested):
Thanks for updates and fixes @mmlb , for the reviewers sake, going ahead could we keep the PRs scope limited

mmlb commented:
> Thanks for updates and fixes @mmlb , for the reviewers sake, going ahead could we keep the PRs scope limited

I'm going to drop new features from this PR and just do the format/lint fixes, aka the cleaning part

mmlb commented:
> Thanks for updates and fixes @mmlb , for the reviewers sake, going ahead could we keep the PRs scope limited

I've trimmed things down and also split the commits a little bit so that they are easier to review. I split out what was previously "automatic fixes + manual fixes" into multiple ones. The first iteration was all together so that each commit would still pass linting so it was quite large. Now a couple of intermediate commits won't pass but makes for smaller diffs.

joelrebel reviewed (approved):
This looks fine to me

jacobweinstock commented on providers/supermicro/firmware_bios_test.go:
```diff
@@ -93,7 +94,7 @@ func Test_setComponentUpdateMisc(t *testing.T) {
 				}
 
-				assert.Nil(t, err)
+				assert.NoError(t, err)
```
The use of `require.NoError` and `assert.NoError` seems arbitrary. When to use which is not clear. Is this something that `golangci-lint` catches? If not, I recommend either creating a doc for when one should be used over the other, or just stick to one.

jacobweinstock commented on internal/redfishwrapper/task.go:
```diff
@@ -97,6 +95,8 @@ func (c *Client) TaskStateActive(state constants.TaskState) (bool, error) {
 		return true, nil
 	case constants.Complete, constants.Failed:
 		return false, nil
+	case constants.PowerCycleHost, constants.Unknown:
+		fallthrough
```
For what is this optimizing? Why is this needed if it's just going to fallthrough to the default? Won't the default case be called regardless of whether this case exists or not?

jacobweinstock replied:
Oh, is this to satisfy a linting complaint?

joelrebel replied:
`constants.PowerCycleHost` and `constants.Unknown` should not be part of the switch case since they are not Redfish task states,
could we have this case removed?

http://redfish.dmtf.org/schemas/DSP2046_2019.4.html#redfish.dmtf.org/schemas/v1/Task.json|details|TaskState

jacobweinstock reviewed (commented):
> Thanks for updates and fixes @mmlb , for the reviewers sake, going ahead could we keep the PRs scope limited

I don't think multiple commits was what @joelrebel was driving at here. Limiting scope would be, for example, creating a PR just for the linting changes, one for the small pedantic differences you prefer, and another for just the proxy support.

joelrebel commented:
Closing this PR as its not being worked on anymore
